File: practices/leetcode/binary_search/69-sqrtx.cpp
// Solution 1: binary search to the right
class Solution {
public:
    int mySqrt(int x) {
        // Find max y such that y^2 = x
        int lo = 0;
        int hi = x;
        while(lo < hi) {
            int dt = hi - lo;
            int mi = lo + dt / 2 + (dt % 2);
            
            if (mi <= x / mi) lo = mi;
            else hi = mi - 1;
        }
        return lo;
    }
};

// Solution 2: binary search to the left
class Solution:
    def mySqrt(self, x: int) -> int:
        lo = 1
        hi = x + 1
        while lo < hi:
            mi = (lo + hi) // 2
            if mi * mi > x: hi = mi
            else: lo = mi + 1
        return hi - 1

File: practices/leetcode/binary_search/69-sqrtx.py
class Solution:
    def mySqrt(self, x):
        # Find maximum y so that y*y <= x
        low = 0
        high = x
        while low < high:
            mid = low + (high - low + 1) // 2
            # mid = (low + high + 1) // 2 = low + (high - low) // 2 + (high - low % 2)
            if mid*mid <= x: low = mid
            else: high = mid - 1
        return low
